### PR TITLE
AuthProxy: Can now login with auth proxy and get a login token

### LINKED ISSUE #17316: Mix Auth Proxy & Grafana login tokens (sessions)
In v6.0 we removed the old session store & session cookies and replaced them with a much more secure auth token scheme. 

This changed how auth proxy worked for some scenarios as the auth tokens are skipped in the auth proxy scenario. 

This issue is track interest in auth proxy scenario that mixes Auth Proxy & Auth tokens. Meaning the auth proxy header is only sent on some auth requests and not all.

### PR DESCRIPTION
**Pull request by torkelo**
Fixes #17316 

Changed so you can login using auth proxy. If you access /login and the your already logged in via auth proxy we now create an auth token so you stay logged in after redirect. 

Added example nginx config to test this scenario. 

Todo:
[x] Unit test for this
[x] Should we have an option for this? 
[x] What other scenarios could cause a /login request and you're already logged in?

### CODE REVIEW

File: pkg/api/login.go [hunk 1]
```diff
@@ -57,18 +57,27 @@ func (hs *HTTPServer) LoginView(c *models.ReqContext) {
 		return
 	}
 
-	if !c.IsSignedIn {
-		c.HTML(200, ViewIndex, viewData)
-		return
-	}
+	// If we already are signed in (Via auth proxy for example)
+	if c.IsSignedIn {
```
Comment (marefr) [inline pkg/api/login.go]: Need to verify that auth proxy is enabled. Would be nice to extract this in similar manner as `tryOAuthAutoLogin`.

File: pkg/api/login.go [hunk 2]
```diff
@@ -57,18 +57,27 @@ func (hs *HTTPServer) LoginView(c *models.ReqContext) {
...
+		// Create login token
+		hs.loginUserWithUser(&models.User{
+			Id:    c.SignedInUser.UserId,
+			Email: c.SignedInUser.Email,
+			Login: c.SignedInUser.Login,
+		}, c)
+
+		if redirectTo, _ := url.QueryUnescape(c.GetCookie("redirect_to")); len(redirectTo) > 0 {
```
Comment (marefr) [inline pkg/api/login.go]: Think this code should still execute last in this method always when logged in and by that having old `if !c.IsSignedIn {` code block before auth proxy enabled and token login. 

Like this:
- Not logged in? Return login page
  - What to do here if auth proxy is enabled? We don't want to render the login form/page right.
- Auth proxy enabled? Auth token login
- Redirect logic

**Review by marefr — CHANGES_REQUESTED** [review]
It's a simple solution, but would suggest putting some more effort into it.

Regarding option/setting to enable this - definitely yes. I don't want to risk as a first step breaking installations due to this.

To me it would make sense to have a specific url like /auth-proxy-login instead of /login to move everything auth proxy related into that. However then we would have to add logic to disable the normal login page and custom redirect rule.

Questions:
Should something happen when auth token is rotated? 
What about using auth proxy and ldap?
What happens when logging out? User redirected to login page and gets logged in again - possible redirect loop problem?

File: conf/defaults.ini [hunk 3]
```diff
@@ -436,6 +436,7 @@ auto_sign_up = true
 ldap_sync_ttl = 60
 whitelist =
 headers =
+grant_login_token = false
```
Comment (marefr) [inline conf/defaults.ini]: Not sure about "grant". How about `enable_login_tokens`?

File: conf/sample.ini [hunk 4]
```diff
@@ -398,6 +398,9 @@
 ;ldap_sync_ttl = 60
 ;whitelist = 192.168.1.1, 192.168.2.1
 ;headers = Email:X-User-Email, Name:X-User-Name
+# If you want to combine auth proxy with the normal login session token & cookie enable setting defined below.
```
Comment (marefr) [inline conf/sample.ini]:
```suggestion
# If you want to combine auth proxy with Grafana login token stored in cookie you can enable setting defined below.
```

File: conf/sample.ini [hunk 5]
```diff
@@ -398,6 +398,9 @@
...
+# It will grant a login token if you supply auth proxy headers on /login request
```
Comment (marefr) [inline conf/sample.ini]: Shouldn't this read if you only provide auth proxy headers for the /login request? Quite hard to explain all of the details in a couple of lines here. Maybe reference to check documentation instead or/and?

File: devenv/docker/blocks/nginx_proxy/nginx_login_only.conf [hunk 6]
```diff
@@ -0,0 +1,42 @@
+events { worker_connections 1024; }
```
Comment (marefr) [inline devenv/docker/blocks/nginx_proxy/nginx_login_only.conf]: Would be good to add this to documentation as an example.

File: docs/sources/auth/auth-proxy.md [hunk 7]
```diff
@@ -36,6 +36,9 @@ whitelist =
 # Optionally define more headers to sync other user attributes
 # Example `headers = Name:X-WEBAUTH-NAME Email:X-WEBAUTH-EMAIL Groups:X-WEBAUTH-GROUPS`
 headers =
+# If you want to combine auth proxy with the normal login session token & cookie enable setting defined below.
+# It will grant a login token if you supply auth proxy headers on /login request
```
Comment (marefr) [inline docs/sources/auth/auth-proxy.md]: Same comment as in default.ini/sample.ini

File: docs/sources/auth/auth-proxy.md [hunk 8]
```diff
@@ -294,3 +297,12 @@ curl -H "X-WEBAUTH-USER: leonard" -H "X-WEBAUTH-GROUPS: lokiteamOnExternalSystem
 With this, the user `leonard` will be automatically placed into the Loki team as part of Grafana authentication.
 
 [Learn more about Team Sync]({{< relref "auth/team-sync.md" >}})
+
+
+## Login token & session cookie
```
Comment (marefr) [inline docs/sources/auth/auth-proxy.md]:
```suggestion
## Login token and session cookie
```

File: docs/sources/auth/auth-proxy.md [hunk 9]
```diff
@@ -294,3 +297,12 @@ curl -H "X-WEBAUTH-USER: leonard" -H "X-WEBAUTH-GROUPS: lokiteamOnExternalSystem
...
+
+If you want to combine auth proxy with the normal login session token & cookie enable the setting `grant_login_token`.
+With that option enabled requests to `/login` that have auth proxy headers added will create a new session token for the
+user (and redirect user). This means that only /login route needs to go through auth proxy. Use settings `login_maximum_inactive_lifetime_days`
```
Comment (marefr) [inline docs/sources/auth/auth-proxy.md]: Should explain something along the lines: In this setup the auth proxy is responsible for authenticating the user, but Grafana will keep track of the session and keep the user logged in why the auth proxy only needs to provide headers for the /login route. 

Would be good to link to `[Login and short-lived tokens](/auth/overview/#login-and-short-lived-tokens) which has all the details about it.

**Review by marefr — CHANGES_REQUESTED** [review]
Works as expected :+1: Some documentation/descriptions left to make this feature understandable, see comments.

File: docs/sources/auth/auth-proxy.md [hunk 10]
```diff
@@ -294,3 +296,13 @@ curl -H "X-WEBAUTH-USER: leonard" -H "X-WEBAUTH-GROUPS: lokiteamOnExternalSystem
 With this, the user `leonard` will be automatically placed into the Loki team as part of Grafana authentication.
 
 [Learn more about Team Sync]({{< relref "auth/team-sync.md" >}})
+
+
+## Login token and session cookie
+
+With `enable_login_token` set to `true` Grafana will, after successfull auth proxy header validation, assign the user
+a login token & cookie. You only have to configure your auth proxy to provide headers for the /login route.
```
Comment (marefr) [inline docs/sources/auth/auth-proxy.md]:
```suggestion
a login token and cookie. You only have to configure your auth proxy to provide headers for the /login route. 
```

File: pkg/setting/setting.go [hunk 11]
```diff
@@ -854,6 +856,8 @@ func (cfg *Cfg) Load(args *CommandLineArgs) error {
 		return err
 	}
 	AuthProxyAutoSignUp = authProxy.Key("auto_sign_up").MustBool(true)
+	AuthProxyEnableLoginToken = authProxy.Key("enable_login_token").MustBool(true)
```
Comment (marefr) [inline pkg/setting/setting.go]:
```suggestion
	AuthProxyEnableLoginToken = authProxy.Key("enable_login_token").MustBool(false)
```

File: pkg/setting/setting.go [hunk 12]
```diff
@@ -854,6 +856,8 @@ func (cfg *Cfg) Load(args *CommandLineArgs) error {
...
+	AuthProxyLDAPSyncTtl = authProxy.Key("ldap_sync_ttl").MustInt()
```
Comment (marefr) [inline pkg/setting/setting.go]: What did this come from?

**Review by marefr — CHANGES_REQUESTED** [review]
Great. Just minor suggestions/comment left

File: pkg/api/login_test.go [hunk 13]
```diff
@@ -136,3 +136,59 @@ func TestLoginOAuthRedirect(t *testing.T) {
...
+		c.SignedInUser = &models.SignedInUser{
+			UserId: 10,
+		}
+		hs.LoginView(c)
+	})
+
+	setting.OAuthService = &setting.OAuther{}
+	setting.OAuthService.OAuthInfos = make(map[string]*setting.OAuthInfo)
+	setting.AuthProxyEnabled = true
+	setting.AuthProxyGrantLoginToken = grantLoginToken
```
Comment (marefr) [inline pkg/api/login_test.go]:
```suggestion
	setting.AuthProxyEnableLoginToken = enableLoginToken
```

File: pkg/api/login_test.go [hunk 14]
```diff
@@ -136,3 +136,59 @@ func TestLoginOAuthRedirect(t *testing.T) {
...
+	location, ok := sc.resp.Header()["Location"]
+	assert.True(t, ok)
+	assert.Equal(t, location[0], "/")
+
+	setCookie, ok := sc.resp.Header()["Set-Cookie"]
+	assert.True(t, ok, "Set-Cookie exists")
+	assert.Equal(t, "grafana_session=; Path=/; Max-Age=0; HttpOnly", setCookie[0])
+}
+
+func setupAuthProxyLoginTest(grantLoginToken bool) *scenarioContext {
```
Comment (marefr) [inline pkg/api/login_test.go]:
```suggestion
func setupAuthProxyLoginTest(enableLoginToken bool) *scenarioContext {
```

File: pkg/api/login.go [hunk 15]
```diff
@@ -57,18 +57,31 @@ func (hs *HTTPServer) LoginView(c *models.ReqContext) {
 		return
 	}
 
-	if !c.IsSignedIn {
-		c.HTML(200, ViewIndex, viewData)
-		return
-	}
+	if c.IsSignedIn {
+		// Assign login token to auth proxy users if grant_login_token = true
```
Comment (marefr) [inline pkg/api/login.go]:
```suggestion
		// Assign login token to auth proxy users if enable_login_token = true
```

**Review by marefr — CHANGES_REQUESTED** [review]
Last suggestions

**Review by marefr — APPROVED** [review]
LGTM

**Comment by gershongad** [comment]
@ torkelo do we need to upgrade grafana (currently we are using 6.2.5) in order to overcome the problem ?
could you please attach the needed nginx and grafana configuration in detail

**Comment by torkelo** [comment]
Yes, this feature requires Grafana v6.5 

example nginx config:
https://github.com/grafana/grafana/blob/master/devenv/docker/blocks/nginx_proxy/nginx_login_only.conf#L31

**Comment by gershongad** [comment]
@torkelo 
Thank you very much for your quick respond,
a. The latest version at grafana site is 6.4.4. Please advise  how to get version 6.5
b. Please advise how to upgrade (we are running centos 7.0)

**Comment by torkelo** [comment]
6.5 is not released yet, beta out this week.

You find nightly builds of 6.5 on Grafana.com

**Comment by duylong** [comment]
Great feature, but how to for alert rules ? I must, each times, use another datasource with a user for basic auth?

**Comment by marefr** [comment]
@duylong not exactly sure what you're referring to. Sounds like you want to put same auth proxy in front of datasource.

**Comment by duylong** [comment]
Sorry for my lack of precision.
Now, I can pass the authentication to our CAS system to access the data (maybe nothing to do with this feature). Before, I had to set up a common user (using the BasicAuth option). After a little test, I notice that I still have to keep this mechanism because the alerts need it (normal, they do not have an authentication session).

**Comment by gershongad** [comment]
@torkelo As we need to integrate grafana in our web site, without asking the user to re-type a password, we want to skip the nginx basic auth username & password dialog. We want to pass the username & password as parameters of the url, or some other solution, so that the user will not have to retype his username & password. Is it possible, and could you please direct us for some php/c etc. code example

**Comment by torkelo** [comment]
Url is not supported, HTTP header is

**Comment by gershongad** [comment]
@torkelo sorry but i do not understand the process, should i display an href like http://user:password@server/grafana/login ??? when i enter such a url to the browser navigation bar it does not reach the destination, or should i generate a curl call from a cgi running on my server, what will be the response for such a curl call ??? i want my users be able to view grafana dashboards. please advise.

**Comment by torkelo** [comment]
you need an nginx proxy that handles the single sign on with your auth system, that proxy can add the required headers for grafana

**Comment by gershongad** [comment]
@torkelo We are already using the nginx proxy, but as i have mentioned the nginx basic auth pops a username & password dialog. We want to be able to skip this dialog

**Comment by torkelo** [comment]
ok, well you have to use some other auth mechanism with nginx then, if you use auth proxy Grafana is not involved in the auth

**Comment by gershongad** [comment]
@torkelo i am trying to use the secure_link nginx auth, but then grafana comes back without a session cookie so that the user is being thrown from his session login. i thought grafana 6.5 should overcome this issue. Please advise

**Comment by gershongad** [comment]
@torkelo i have also tried proxy_set_header Authorization base64(user:password). in case it is hard coded in the nginx.conf file like proxy_set_header Authorization "Basic YWRtaW46VhN12Ud1a284MQ==" it works fine, but when i pass the "YWRtaW46VhN12Ud1a284MQ==" as a url parameter, grafana will come back without a cookie

**Comment by sfitts** [comment]
We proxy access to Grafana through our server since it is where we maintain identities.  Conceptually however is it very similar.  We proxy grafana at `/grafana/`.  When we see `/grafana/login` we determine who the user is and add the appropriate header for Grafana.  At that point Grafana replies with a `FOUND` redirect to `/grafana/`.  Pre-6.x this reply also included the session cookie which the browser would then attach to all subsequent requests.  For any URI that isn't `/grafana/login` we just pass the request directly from the browser to Grafana and since it has the session cookie on it, Grafana knew who the user was (we don't add the header in those cases).

However, at this point, running 6.5.2 things aren't working.  The first part is the same, we proxy login, add the header and get the redirect.  However, there is no cookie (which I guess is intentional) and I don't see anything in the redirect which is going to help Grafana maintain the user's identity on subsequent requests.  I assumed that this fix was going to provide that, so clearly I'm missing something.

At this point, how is this supposed to work?  Assuming the header is only added to the `/login` request, how is the identity established for all subsequent requests?  TIA for any guidance.

**Comment by sfitts** [comment]
OK -- found our issue.  We need to set `enable_login_token` to `true`.

**Comment by gershongad** [comment]
@sfitts could you please attach the nginx.conf & grafana.ini files that you are using ?

**Comment by sfitts** [comment]
As I mentioned, we're not using Nginx (we proxy the requests through a proprietary server).  However, the nginx config provided by @torkelo is equivalent to what we are doing.  The relevant section of the grafana.ini looks like:

```ini
[auth.proxy]
enabled = true
header_name = X-WEBAUTH-USER
header_property = username
auto_sign_up = false
enable_login_token = true
```

**Comment by gershongad** [comment]
@sfitts i works, but i want to skip the username & password pop up, since the user already authenticated in my site. Do you know any way this can be achived ?

**Comment by sfitts** [comment]
Unfortunately I'm not really an Nginx expert (since we don't use it), so I'm not sure.

**Comment by gershongad** [comment]
@torkelo we like grafana very much, but we can't imagine that our users will have to go through a authentication process to view a grafana board, since they have been already authenticated when logged in to our web site. As i understand the nginx basic auth mechanism saves something like a cookie after the user has entered name & password, and in the next accesses the $remote_user is set to the user-name identified by the cookie. Is there another nginx authentication method which works in a similar way, we want to pass somehow the name&password as parameters (might be encrypted) to the grafana/login and imitate the NGINX basic auth mechanism without poping the auth dialog. Is it possible ? maybe we need to write some callback program that will imitate the basic auth mechanism. Please advise in details how this should be done, i think it is needed by many enthusiastic grafana developers.

**Comment by aknuds1** [comment]
@gershongad Don't have an answer for your question, but have you tried asking in the [Grafana community](https://community.grafana.com/)?

**Pull request merged**